kernels/rv64gv/3/bli_gemm_rv64gv_asm_s16x4.c

```diff
@@ -0,0 +1,123 @@
+      from this software without specific prior written permission.
+
+   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+   "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+   A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+   HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+   DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+   THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+   (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+   OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+*/
+#include "blis.h"
+
+extern
+void bli_sgemm_rv64gv_ker_16x4
+     (
+       uint64_t            k,
+       float*     restrict alpha,
+       float*     restrict a,
+       float*     restrict b,
+       float*     restrict beta,
+       float*     restrict c, uint64_t rs_c, uint64_t cs_c
+     );
+
+
+static uint64_t num_fp32_per_vector() {
+    uint64_t velem = 0;
+    __asm__ volatile
+    (
+    " li a5, 256                         \n\t"
+    " vsetvli a4, a5, e32, m1, ta, ma    \n\t"
+    " mv %[velem], a4                    \n\t"
+    : [velem] "=r" (velem)
+    :
+    : "a4", "a5"
+    );
+    return velem;
+}
+
+void bli_sgemm_rv64gv_asm_16x4
+     (
+       dim_t               m,
+       dim_t               n,
+       dim_t               k,
+       float*     restrict alpha,
+       float*     restrict a,
+       float*     restrict b,
+       float*     restrict beta,
+       float*     restrict c, inc_t rs_c0, inc_t cs_c0,
+       auxinfo_t*          data,
+       cntx_t*             cntx
+     )
+{
+    // Use local copy in case dim_t has a different size than expected in the assembly kernel
+    uint64_t _k     = k;
+    uint64_t rs_c   = rs_c0;
+    uint64_t cs_c   = cs_c0;
+
+    GEMM_UKR_SETUP_CT( s, 16, 4, false );
+
+    // The assembly kernel assumes that the vector length is 128 bits.
+    // For now, fall back to generic implementation
+    // if VLEN != 128.
+    uint64_t velem = num_fp32_per_vector();
+
+    if (velem == 4) {
+        bli_sgemm_rv64gv_ker_16x4(_k, alpha, a, b, beta, c, rs_c, cs_c);
+    }
+    else {
+        float ab[16 * 4];
+        for (dim_t i = 0; i < 16 * 4; i++) {
+            ab[i] = 0.0;
+        }
+
+        for (dim_t l = 0; l < k; ++l) {
+            for (dim_t j = 0; j < 4; j++) {
+                for (dim_t i = 0; i < 16; ++i) {
+                    float bj = b[j];
+                    float ai = a[i];
+                    ab[i + j * 16] += ai * bj;
+                }
+            }
+            a += 16;
+            b += 4;
+        }
+
+
+        for (dim_t i = 0; i < 16 * 4; i++) {
+            ab[i] = ab[i] * (*alpha);
```

Comment:
If `alpha == 0`, `ab[i]` should be zeroed regardless of `a[i]` and `b[i]`. GEMM conventions say that `alpha == 0` or `beta == 0` should short-circuit multiplications to `0` even in the presence of `NaN` entries.

Reply:
`alpha == 0` is handled at a much higher level. The micro-kernels are only responsible for handling `beta == 0` and `k == 0`.

Reply:
No objection here, just want to point out that while `alpha == 0` is handled higher up, `k == 0` must still be handled properly in the microkernel. (This may have gone without saying.)

Reply:
Long-since handled correctly.